### PR TITLE
added Grid.hpp

### PR DESCRIPTION
**Pull request by swagween**
added Grid.hpp with a Grid class made up of Cell structs.

### CODE REVIEW

File: src/Grid.hpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
+//
+//  Grid.hpp
+//  Game of Life
+//
+//  Created by Alex Frasca on 01/05/2023.
+//
+#pragma once
+#ifndef Grid_hpp
```
Comment (karnkaul) [inline src/Grid.hpp]: Either `#pragma once` or a header guard, not both

File: src/Grid.hpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+#include <SFML/Graphics.hpp>
+#include <vector>
+#include <cmath>
+#include <functional>
+
+namespace conway {
+
+//several values so we can draw nice colors as the cells die instead of just on/off
+enum CELL_STATE {
```
Comment (karnkaul) [inline src/Grid.hpp]: User strongly typed enums: `enum class`

File: src/Grid.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    CELL_OFF = 0,
+    CELL_LOW = 1,
+    CELL_MEDIUM = 2,
+    CELL_HIGH = 3,
+    CELL_ON = 4
+};
+const float DEFAULT_SPACING = 16.0;
```
Comment (karnkaul) [inline src/Grid.hpp]: `constexpr` for such constants

File: src/Grid.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+struct Cell {
+    
+    Cell() {};
```
Comment (karnkaul) [inline src/Grid.hpp]: This struct should follow the Rule of Zero, it doesn't need custom constructors

File: src/Grid.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    // only for when the size or spacing of the grid is changed, not for each tick
+    void updatePosition();
+    
+    // Cell member vars
+    sf::Vector2<int> index;
+    sf::Vector2<float> position;
+    // for drawing the cell
+    sf::ConvexShape drawable;
+    
+    float spacing;
```
Comment (karnkaul) [inline src/Grid.hpp]: Variables should be initialized when they are defined (via `{}`).

File: src/Grid.hpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    Grid(uint32_t width, uint32_t height);
+    
+    uint32_t grid_width;
+    uint32_t grid_height;
+    
+    // how far apart are the grid elements?
+    float spacing;
+    
+    //the cells of the grid
+    std::vector<Cell> cells;
```
Comment (karnkaul) [inline src/Grid.hpp]: Should be private to the class, otherwise it can't guarantee maintenance of associated invariants

File: src/Grid.hpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    void random_init();
+    void initialize();
+    void update();
+    
+    //for Game of Life rules and helper functions
+    void tick_cells();
+    std::vector<std::reference_wrapper<Cell> > get_neighbors(Cell& current_cell);
+    
+    //grid init functions
+    void setSpacing(float spc);
```
Comment (karnkaul) [inline src/Grid.hpp]: Inconsistent casing: snake vs camel

**Review by karnkaul — CHANGES_REQUESTED** [review]
Where are the member functions defined? There's no cpp file

File: src/Grid.hpp (hunk 1 above)
Comment (swagween) [reply src/Grid.hpp]: fixed.

File: src/Grid.hpp (hunk 2 above)
Comment (swagween) [reply src/Grid.hpp]: done.

File: src/Grid.hpp (hunk 3 above)
Comment (swagween) [reply src/Grid.hpp]: done.

File: src/Grid.hpp (hunk 4 above)
Comment (swagween) [reply src/Grid.hpp]: I just removed the `<Cell() {}>` here. It can still have a constructor with parameters, right?

File: src/Grid.hpp (hunk 5 above)
Comment (swagween) [reply src/Grid.hpp]: done.

File: src/Grid.hpp (hunk 6 above)
Comment (swagween) [reply src/Grid.hpp]: just the vector? or should other vars be private as well?

File: src/Grid.hpp (hunk 7 above)
Comment (swagween) [reply src/Grid.hpp]: fixed.

File: src/Grid.hpp (hunk 4 above)
Comment (karnkaul) [reply src/Grid.hpp]: It doesn't need any, users can use braced or designated init to initialize specific members.

File: src/Grid.hpp (hunk 6 above)
Comment (karnkaul) [reply src/Grid.hpp]: Haven't gone through the logic to be able to answer that accurately, but in general anything that if poked at from outside the class could mess up the state of the object, is an invariant the class needs to maintain, and thus cannot be public.

**Pull request merged**